File: TC Programa 1/SigmaUniverso.c
#include<stdio.h>
#include<stdlib.h>
#include<math.h>
#include<time.h>

int menu();
void sigma(int, FILE*);
void binario(int,int, FILE*);

int main(){
    srand(time(NULL));
    FILE *doc;
    doc=fopen("Universo.txt","w"); //inicia con un archivo en blanco
  //doc=fopen("Universo.txt","a"); 
    int op;
    int n;
    while(1){
        op=menu();
        switch(op){
        case 1:
            printf("\nIngrese el valor de n: ");
            scanf("%d",&n);
            printf("Generando el universo de n = %d...\n\n",n);
            fprintf(doc,"\tUniverso de n = %d\n\nU = {e",n);
            sigma(n,doc);
            fputs("}\n\n",doc);
            printf("Se genero correctamente el universo de n = %d\n\n",n);
            break;
        case 2:
            n=rand()%1001;
            printf("Generando el universo de n = %d...\n\n",n);
            fprintf(doc,"\tUniverso de n = %d\n\nU = {e",n);
            sigma(n,doc);
            fputs("}\n\n",doc);
            printf("Se genero correctamente el universo de n = %d\n\n",n);
            break;
        case 3:
            printf("\nHasta luego :D");
            exit(0);
            break;
        default:
            break;
        }
    }
    return 0;
}

void binario(int n, int elem, FILE *doc){
    if(n==0){
        while(elem>0){
            fputc('0',doc);
            elem--;
        }
        return;
    }
    binario(n/2,elem-1,doc);
    fprintf(doc,"%d",n%2);
}

void sigma(int n, FILE *doc){
    if(n==0){
        return;
    }
    sigma(n-1,doc);
    //lo que genera sigma de n
    int i;
    for(i=0;i<pow(2,n);i++){
        fputs(", ",doc);
        binario(i, n, doc);
    }
}

int menu(){
    int respuesta = 0;
	printf("\n\t\t    Menu\n============================================\n");
	printf("1. Generar uinverso n\n");
	printf("2. Generar univerrso con n aleatorio\n");
	printf("3. Salir\n\nSeleccione una opcion: ");
	scanf("%d",&respuesta);
	while(respuesta>3 || respuesta<1){
        fflush(stdin);
		printf("\n\t> Error <\n\nSeleccione una opcion valida: ");
		scanf("%d",&respuesta);
	}return respuesta;
}

File: TC Programa 1/SigmaUniverso.py
import matplotlib.pyplot as plt
import random
import math

def menu():
    print('\n\t\t    Menu\n============================================')
    print('1. Generar uinverso n')
    print('2. Generar univerrso con n aleatorio')
    print('3. Generar gráficas para el universo anterior')
    print('4. Salir\n')
    op=int(input('Seleccione una opcion: '))
    while(op>4 or op<1):
        print('\n\t> Error <\n')
        op=int(input('Seleccione una opcion valida: '))
    return op

def binario(n, elem, conteo):
    if(n==0):
        while(elem>0):
            doc.write('0')
            elem-=1
        return conteo
    conteo=binario(n//2,elem-1,conteo)
    if(n%2==1):
        conteo+=1
    doc.write(str(n%2))
    return conteo

def sigma(n):
    if(n==0):
        return
    sigma(n-1)
    for i in range (2**n):
        doc.write(', ')
        elementos.append(n)
        unos.append(binario(i,n,0))
        cad.append(cad[-1]+1)

doc=open('Universo.txt','w')

elementos=[]    # Número de elementos por cadena
unos=[]         # Número de unos por cadena
cad=[]          # Número de cadenas distintas

while True:
    op=menu()
    if op==1:       # Generar uinverso n
        unos.clear()
        elementos.clear()
        cad.clear()
        n=int(input('\nIngrese el valor de n: '))
        print('\nGenerando el universo de n =',str(n)+'...\n')
        doc.write('>>> Universo de n = '+str(n)+'\n\nU = {e')
        cad.append(0)
        unos.append(0)
        elementos.append(0)
        sigma(n)
        doc.write('}\n\n')
        print('Se genero correctamente el universo de n =',n)

    elif op==2:     # Generar univerrso con n aleatorio
        cad.clear()
        unos.clear()
        elementos.clear()
        n=random.randint(1,1000)
        print('\nGenerando el universo de n =',str(n)+'...\n')
        doc.write('>>> Universo de n = '+str(n)+'\n\nU = {e')
        cad.append(0)
        unos.append(0)
        elementos.append(0)
        sigma(n)
        doc.write('}\n\n')
        print('Se genero correctamente el universo de n =',n)

    elif op==3:     # Generar gráficas para en universo anterior
        if len(cad)==0:
            print('\nNo se ha generado ningun universo, por favor seleccione otra opcion')
        else:
            print('\nSe generaron las graficas para el universo n =',str(n)+', cierre las ventanas para continuar')
            
            logE=[]

            for i in elementos:
                if i==0:
                    logE.append(0)
                else:
                    logE.append(math.log(i, 10))

            plt.title('Elementos por cadena universo n = '+str(n))    
            plt.xlabel('Cadena')
            plt.ylabel('Elementos')
            plt.plot(cad, elementos, label='Elementos')
            plt.plot(cad, logE, label='Log Elementos')
            plt.legend(loc='right')
            plt.show()

            logU=[]
            
            for i in unos:
                if i==0:
                    logU.append(0)
                else:
                    logU.append(math.log(i, 10))            

            plt.title('Unos por cadena universo n = '+str(n))    
            plt.xlabel('Cadena')
            plt.ylabel('Unos')
            plt.plot(cad, unos, label='Unos')
            plt.plot(cad, logU, label='Log Unos')
            
            plt.legend(loc='upper left')
            plt.show()

    elif op==4:     # Salir
        print('\nHasta luego :D\n')
        break
doc.close()